packages/editor/src/store/actions.js

```diff
@@ -59,8 +119,12 @@ export function* setupEditor( post, edits, template ) {
 		content = post.content.raw;
 	}
 
+	// Parse content blocks
 	let blocks = parse( content );
 
+	// Augment with post attributes
+	blocks = mapBlocks( blocks, ( block ) => metaSource.synchronize( block, post.meta ) );
```

Comment:
It kinda makes me wonder why we're authoring these sources as if they're some generic interface, when ultimately we call it very explicitly and with a very use-case-specific set of arguments.

Reply:
Agreed, right now it's not the perform generic API, it's a first attempt. I think ideally I'd like something composed of two APIs: one API to the list of changes (edits) to apply to the editable elements (post object and ultimately, template object as well)

```
const edits = applyBlockAttributesEdits( attributesConfig, block.attributes );
// edits here is somehting like { post: { meta: { key: value } }
```

the second API is how to update block attributes based on the editable objects (post and tempalte)

```
const block = fillSourceAttributes( block, post, template )
```

Still not perfect but you get the idea.

Reply:
A thought: What if we treat the source synchronization just as any other store control?

That way, they could yield using existing `select` controls to retrieve whatever store data they need (like here, the post meta).

```js
for ( let block of blocks ) {
	for ( let source of sources ) {
		block = yield* source.sync( block );
	}
}

// ...

function* sync( block ) {
	const meta = yield select( 'core/editor', 'getEditedPostAttribute', 'meta' );

	// ...
}
```

Reply:
This does make sense, we'll be calling the same selector over and over for each block though, so I wonder about performance.

Reply:
> This does make sense, we'll be calling the same selector over and over for each block though, so I wonder about performance.

I was thinking about it some more afterward, and at least as far as the looping goes, I was thinking the sources could be defined as named exports from a folder corresponding to the attribute source:

```js
// src/store/sources/index.js
export { meta } from './meta';

// src/store/actions.js
import * as sources from './sources';

for ( let i = 0; i < blocks.length; i++ ) {
	const block = blocks[ i ];
	for ( const schema of getBlockType( block.name ).attributes ) {
		if ( sources[ schema.type ] ) {
			blocks[ i ] = yield* source.sync( blocks[ i ] );
		}
	}
}
```

To your point, I'd wonder if we could memoize (per sync pass) results from some source "common values" which aren't block specific (like meta).

```js
const sourceValues = {};
for ( let i = 0; i < blocks.length; i++ ) {
	const block = blocks[ i ];
	for ( const schema of getBlockType( block.name ).attributes ) {
		if ( ! sources[ schema.type ] ) {
			continue;
		}

		if ( ! sourceValues.hasOwnProperty( schema.type ) ) {
			sourceValues[ schema.type ] = yield* source.getValues();
		}

		blocks[ i ] = yield* source.sync( blocks[ i ], sourceValues[ schema.type ] );
	}
}
```

In the above, there's still a potential performance concern from iterating _every attribute_ for _every block_. Something like #16316 benefits from the fact that this is consolidated into the initial parse phase.

Reply:
yeah, let's try that maybe? this API can remain private for a while.